### ISSUE TITLE
About YOLO11 Confusion Matrix

### ISSUE THREAD

**Issue by MIGANEKAWAKO** [ultralytics/ultralytics#20898]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

There is a question bothers me a lot, my mAP50 is high but the confusion matrix looks bad, because there are too many False Positive(FP). When I change the **conf** and **iou** in the validation, the FP in confusion matrix are reduced but mAP50 also dropped. How should I do to reduce the FP and remain the mAP50?

There are the results:
The original confusion matrix with **mAP50=0.873**:
![Image](https://github.com/user-attachments/assets/fdec547a-bb3f-4568-9d62-ed400a4ef07d)
Set **conf=0.35** and **iou=0.5** with **mAP50=0.865**:
![Image](https://github.com/user-attachments/assets/03cf6fea-797b-4d37-9538-bb0d835b83aa)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @MIGANEKAWAKO, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help you get started quickly. An Ultralytics engineer will also review your question and assist you further soon. Thank you for your patience and for being part of the Ultralytics community! 💡

**Comment by Y-T-G** [comment]
mAP is not calculated after applying a confidence threshold. If you apply a confidence threshold before calculating mAP, then the mAP is wrong.

The confusion matrix is based on confidence threshold 0.25 by default. It's applied internally.

**Comment by MIGANEKAWAKO** [comment]
> mAP is not calculated after applying a confidence threshold. If you apply a confidence threshold before calculating mAP, then the mAP is wrong.应用置信度阈值后不计算 mAP。如果你在计算 mAP 之前应用置信度阈值，那么 mAP 是错误的。
> 
> The confusion matrix is based on confidence threshold 0.25 by default. It's applied internally.默认情况下，混淆矩阵基于置信度阈值 0.25。它会在内部应用。

Thanks for your reply! This is my code, I do this when I want to validate my trained model on a test set. When I want to change conf and iou, should I apply it during training?
`model = YOLO('runs/detect/train/weights/best.pt')
model.val(data="GreenHouseTomato.yaml", device=0, split='test', conf=0.3, iou=0.5, save_json=True, batch=16)`

**Comment by glenn-jocher** [comment]
The behavior you're seeing is expected and not contradictory. mAP measures performance across all confidence levels, while the confusion matrix shows results at your specified confidence threshold.

For proper model evaluation, use the default validation parameters: `model.val(data="GreenHouseTomato.yaml", split='test')`. The conf and iou parameters in validation are meant for tuning your deployment threshold, not for training adjustments.

If you need fewer false positives in your application, experiment with higher confidence thresholds during inference/prediction, but keep validation parameters at defaults to get accurate performance metrics. Training-time parameters won't solve this issue since it's about the operating point you choose for deployment.

**Comment by MIGANEKAWAKO** [comment]
> The behavior you're seeing is expected and not contradictory. mAP measures performance across all confidence levels, while the confusion matrix shows results at your specified confidence threshold.
> 
> For proper model evaluation, use the default validation parameters: `model.val(data="GreenHouseTomato.yaml", split='test')`. The conf and iou parameters in validation are meant for tuning your deployment threshold, not for training adjustments.
> 
> If you need fewer false positives in your application, experiment with higher confidence thresholds during inference/prediction, but keep validation parameters at defaults to get accurate performance metrics. Training-time parameters won't solve this issue since it's about the operating point you choose for deployment.

Thanks for your clear explanation! That is, I just need to adjust the confidence level?

**Comment by glenn-jocher** [comment]
Yes, exactly! For deployment, adjust the confidence threshold during inference to reduce false positives to your desired level. Keep validation at defaults for accurate performance metrics, then tune confidence in your actual application based on your tolerance for FPs vs missed detections.

**Comment by MIGANEKAWAKO** [comment]
> Yes, exactly! For deployment, adjust the confidence threshold during inference to reduce false positives to your desired level. Keep validation at defaults for accurate performance metrics, then tune confidence in your actual application based on your tolerance for FPs vs missed detections.

Thank you so much! Your answer was very helpful to me!

**Comment by glenn-jocher** [comment]
You're very welcome! Glad I could help clarify the difference between validation metrics and deployment thresholds. Feel free to open a new issue if you have any other questions. Happy detecting! 🚀
